sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1902,218 @@ def transform(self, X):
+
+    def __init__(self, inclusion='all'):
+        self.inclusion = inclusion
+
+    @staticmethod
+    def _check_params(inclusion=None):
+        if (inclusion is not None and inclusion != 'all' and
+                not CountFeaturizer._valid_data_type(inclusion)):
+            raise ValueError("Illegal data type in inclusion")
+
+    @staticmethod
+    def _valid_data_type(type_check):
+        return type(type_check) == np.ndarray or type(type_check) == list
+
+    def fit(self, X, y=None):
+        """Fits the CountFeaturizer to X, y
+
+        Stores the counts for each example X, conditional on y
+        Both X and y must be appropriately reshaped to a 2D list
+
+        Parameters
+        ----------
+        X : array
+            The data set to learn the counts from, conditional to 'y'
+            X must not be 1 dimensional
+
+        y : array-like, optional
+            If provided, a separate column is output for each value of 'y',
+            counting the occurences of 'X' conditioned on that 'y' value
+
+        Returns
+        -------
+        self
+        """
+
+        CountFeaturizer._check_params(inclusion=self.inclusion)
+
+        if y is not None:
+            X, y = check_X_y(X, y, multi_output=True)
+            if len(y.shape) == 1:
+                self.col_num_Y_ = 1
+            else:
+                self.col_num_Y_ = len(y[0])
+        else:
+            X = check_array(X)
+            self.col_num_Y_ = 1
+
+        len_data = len(X)
+        self.col_num_X_ = len(X[0])
+        self.count_cache_ = defaultdict(
+            lambda: defaultdict(lambda: defaultdict(int)))
+        self.y_set_ = [set() for i in range(self.col_num_Y_)]
+
+        if type(self.inclusion) == str and self.inclusion == 'all':
+            inclusion_used = np.array(range(self.col_num_X_))
+        elif self.inclusion is None:
+            inclusion_used = np.array([])
+        elif CountFeaturizer._valid_data_type(self.inclusion):
+            inclusion_used = np.array(self.inclusion)
+
+        if y is not None:
+            for i in range(len_data):
+                for j in range(self.col_num_Y_):
+                    X_key = tuple(X[i].take(inclusion_used))
+                    if len(y.shape) == 1:
+                        # if y is 1D, y[i] is a scalar, which is not iterable
+                        y_key = tuple([y[i]])
+                    else:
+                        y_key = tuple(y[i].take([j]))
+                    self.count_cache_[X_key][j][y_key] += 1
+                    self.y_set_[j].add(y_key)
+        else:
+            self.y_set_[0].add(0)
+            for i in range(len_data):
+                X_key = tuple(X[i].take(inclusion_used))
+                self.count_cache_[X_key][0][0] += 1
+        self.y_set_ = [list(enumerate(sorted(ys))) for ys in self.y_set_]
+        # freeze the dicts for pickling
+        self.count_cache_.default_factory = None
+        for cc_inner_dict in self.count_cache_.values():
```

Comment:
What do you think of this part here? Right now because of the way count_cache_ is coded in terms of default dict, I have to iterate though all dicts inside and freeze them one by one.
```
        self.count_cache_ = defaultdict(
             lambda: defaultdict(lambda: defaultdict(int)))
```

This would not be a problem if I were to code it as  
```
        self.count_cache_ = defaultdict(get_first_dict)

def get_first_dict():
        return defaultdict(get_second_dict())

# outside of CountFeaturizer class declaration
def get_second_dict():
        return defaultdict(int)
```